### PR TITLE
[@kadena-clli] Add devnet simulate command (Feat)

### PR DESCRIPTION
**Pull request by nil-amrutlal-dept**
In the PR, we find the simulate feature integrated in `kadena-cli` under the `devnet` command.
The simulate feature is responsible for simulating traffic in the devnet. Alongside the simulation, some auxiliary files implement the different transfer types and utils functions. 
The simulate feature also produces a log file. 

Note: some additional files are included in the PR due to lint and style fixes 

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1205886760574941

### CODE REVIEW

**Review by MRVDH — APPROVED** [review]

File: packages/tools/kadena-cli/src/account/commands/accountFund.ts [hunk 1]
```diff
@@ -8,66 +8,62 @@ import { FAUCET_CONSTANTS } from '../../constants/faucet.js';
...
-  [
-    globalOptions.accountName(),
-    globalOptions.amount(),
-    globalOptions.fungible(),
-    globalOptions.network(),
-    globalOptions.chainId(),
-  ],
-  async (config) => {
-    debug('account-fund:action')({ config });
+export const fundCommand: (program: Command, version: string) => void =
```
Comment (nillo) [inline packages/tools/kadena-cli/src/account/commands/accountFund.ts:11]: Please align ur ide :) and remove all the files that are not related to your actual changes from this PR

File: packages/tools/kadena-cli/src/devnet/utils/simulation/file.ts [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+  };
+  const directory = path.join(`${folder}/`);
+  createDir(directory);
+  const filepath = path.join(directory, filename);
+  const headers = `${Object.keys(exampleData).join(',')}\n`;
+  fs.writeFileSync(filepath, headers);
+  return filepath;
+}
+
+export function appendToFile(
```
Comment (nillo) [inline packages/tools/kadena-cli/src/devnet/utils/simulation/file.ts:49]: @barthuijgen throughout this whole file helper you want it to make use of your service so it can use memoryfs?

File: packages/tools/kadena-cli/src/utils/globalOptions.ts [hunk 3]
```diff
@@ -195,6 +196,64 @@ export const globalOptions = {
       'Version of the kadena/devnet Docker image to use (e.g. "latest")',
     ),
   }),
+  //Simulate
```
Comment (nillo) [inline packages/tools/kadena-cli/src/utils/globalOptions.ts:199]: please take a look at this comment, since i have the same comment here for this code: https://github.com/kadena-community/kadena.js/pull/1476/files#r1453533628

**Review by nillo — CHANGES_REQUESTED** [review]
@alber70g same question as with other PR, i have looked at the implementation, but other then that, its usecase / usage is something i have not enough input on please review

File: packages/tools/kadena-cli/src/account/commands/accountFund.ts (hunk 1 above)
Comment (nil-amrutlal-dept) [reply packages/tools/kadena-cli/src/account/commands/accountFund.ts]: Fixed, removed the non-related files

File: packages/tools/kadena-cli/src/utils/globalOptions.ts (hunk 3 above)
Comment (nil-amrutlal-dept) [reply packages/tools/kadena-cli/src/utils/globalOptions.ts:199]: Fixed

File: packages/tools/kadena-cli/src/devnet/utils/simulation/file.ts (hunk 2 above)
Comment (barthuijgen) [reply packages/tools/kadena-cli/src/devnet/utils/simulation/file.ts:49]: Yes definitely! This does seem like code that would be good to add unit test for. 

Please make use of the `services.filesystem` implementation, it does not have appendFile yet, but you can add it to the service, send me a message if it is unclear how to do that.

**Review by alber70g — CHANGES_REQUESTED** [review]

File: packages/tools/kadena-cli/README.md [hunk 4]
```diff
@@ -79,6 +79,18 @@ brew install kda-cli
...
+To install the executable from this repo:
+
+```sh
+pnpm install
+pnpm build --filter @kadena/kadena-cli
+# if you are using NVM, you should have this environment variable available
+ln -s $(pwd)/bin/kadena-cli.js $NVM_BIN/kda
+# if not, you can replace $NVM_BIN to any path you have added in your $PATH
+```
+
```
[the remark this answers is not in the document]
Comment (nil-amrutlal-dept) [reply packages/tools/kadena-cli/README.md:93]: Yep, so I had to do this to be able to use the `kda` command from any directory on my terminal. In my case (not using brew) I had to create a symlink to the bin file, which in my case (using NVM), is `$NVM_BIN/kda`. 
`$NVM_BIN`: environment variable which specifies my bin directory. I can rewrite it to make it more clear. 
On my case `pnpm link` and `pnpm install -g` did not work

File: packages/tools/kadena-cli/src/devnet/utils/simulation/coin/crosschain-transfer.ts [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+    );
+    gasPayer = sender00;
+  }
+
+  if (!gasPayer.keys.map((key) => key.secretKey)) {
+    console.log(
+      `Gas payer ${gasPayer.account} does not have a secret key; using sender00 as gas payer`,
+    );
+    gasPayer = sender00;
+  }
```
[the remark this answers is not in the document]
Comment (nillo) [reply packages/tools/kadena-cli/src/devnet/utils/simulation/coin/crosschain-transfer.ts:34]: i think these validations should not be part of the options since hes comparing the result of 2 totally we do this also in other places, and it would complicate the option logic

File: packages/tools/kadena-cli/src/utils/globalOptions.ts [hunk 6]
```diff
@@ -195,6 +196,64 @@ export const globalOptions = {
...
+  }),
+  simulateLogFolder: createOption({
+    key: 'logFolder' as const,
+    prompt: simulatePrompts.simulateLogFolderPrompt,
+    validation: z.string(),
+    option: new Option(
+      '-l, --log-folder <logFolder>',
+      'Specify the directory where the log file will be generated.',
+    ),
+  }),
```
[the remark this answers is not in the document]
Comment (nil-amrutlal-dept) [reply packages/tools/kadena-cli/src/utils/globalOptions.ts:255]: Having also this in mind: https://github.com/kadena-community/kadena.js/pull/1476#discussion_r1453533628

What should be the go-to behavior? The first option should be a one letter flag when possible (since the second already states what it is) or should we keep one letter flags for booleans and common options?

File: packages/tools/kadena-cli/src/account/commands/accountFund.ts (hunk 1 above)
Comment (nil-amrutlal-dept) [reply packages/tools/kadena-cli/src/account/commands/accountFund.ts]: Note: these files were auto formatted using `pnpm format:src` so I am not sure if the problem is in my IDE

File: packages/tools/kadena-cli/src/devnet/utils/simulation/file.ts (hunk 2 above)
Comment (nil-amrutlal-dept) [reply packages/tools/kadena-cli/src/devnet/utils/simulation/file.ts]: Fixed

File: packages/tools/kadena-cli/README.md (hunk 4 above)
[the remark this answers is not in the document]
Comment (nil-amrutlal-dept) [reply packages/tools/kadena-cli/README.md]: Ended up installing a different version manager that makes use of a shared bin directory. Managed to install `kadena` globally by doing `pnpm link -g`

File: packages/tools/kadena-cli/src/devnet/commands/devnetSimulation.ts [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+      if (!containerIsRunning(config.devnetConfig.name)) {
+        console.log('Devnet is not running. Please run the devnet first.');
+        return;
+      }
+
+      await simulateCoin({
+        network: {
+          id: 'fast-development',
+          host: `http://localhost:${config.devnetConfig.port}`,
+        },
```
Comment (alber70g) [inline packages/tools/kadena-cli/src/devnet/commands/devnetSimulation.ts:38]: can we use `globalOptions.network()` here?

File: packages/tools/kadena-cli/src/devnet/utils/docker.ts [hunk 8]
```diff
@@ -141,3 +141,13 @@ export function updateDevnet(version?: string): void {
...
+
+export function containerIsRunning(name: string): boolean {
+  const runningContainers = execSync(
+    'docker ps --filter ancestor=kadena/devnet --format "{{.Names}}"',
+  )
+    .toString()
+    .trim()
+    .split('\n');
+  return runningContainers.includes(name);
+}
```
Comment (alber70g) [inline packages/tools/kadena-cli/src/devnet/utils/docker.ts:153]: When using `ancestor=kadena/devnet`, argument `name` becomes obsolete and you can check whether the amount if lines returned by `docker ps .... ` > 0, right?

File: packages/tools/kadena-cli/src/devnet/utils/simulation/coin/crosschain-transfer.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (alber70g) [reply packages/tools/kadena-cli/src/devnet/utils/simulation/coin/crosschain-transfer.ts:33]: Can we add `sender00` as a predefined global account?

File: packages/tools/kadena-cli/src/devnet/utils/simulation/coin/safe-transfer.ts [hunk 9]
```diff
@@ -0,0 +1,92 @@
...
+  sender,
+  amount,
+}: {
+  network: { host: string; id: string };
+  receiver: IAccount;
+  chainId: ChainId;
+  sender: IAccount;
+  amount: number;
+}): Promise<ICommandResult> {
+  const extraAmount = new PactNumber('0.001').toPactDecimal();
```
Comment (alber70g) [inline packages/tools/kadena-cli/src/devnet/utils/simulation/coin/safe-transfer.ts:22]: Can we use the smallest amount possible? 1*10^-8

**Review by alber70g — APPROVED** [review]

File: packages/tools/kadena-cli/src/devnet/utils/docker.ts (hunk 8 above)
Comment (nil-amrutlal-dept) [reply packages/tools/kadena-cli/src/devnet/utils/docker.ts:153]: https://github.com/kadena-community/kadena.js/pull/1469#discussion_r1457389453

File: packages/tools/kadena-cli/src/devnet/commands/devnetSimulation.ts (hunk 7 above)
Comment (nil-amrutlal-dept) [reply packages/tools/kadena-cli/src/devnet/commands/devnetSimulation.ts:38]: We can indeed use it. However, since I already prompt the user to choose a `devnet` I don't know if it's necessary to have both of the options. The reason for the `devnet` prompt instead of `network`, is that we can validate if the selected devnet configuration is running (by verifying the running containers against the devnet name)

File: packages/tools/kadena-cli/src/devnet/utils/docker.ts (hunk 8 above)
Comment (nil-amrutlal-dept) [reply packages/tools/kadena-cli/src/devnet/utils/docker.ts:153]: The reason for verifying the name is that we can make sure the chosen devnet config is indeed the one running. This way we not only validate that a devnet is indeed running but also that the devnet configuration chosen is the correct one. This is useful because we can therefore use the port in that config without the need for an additional promt

File: packages/tools/kadena-cli/src/utils/globalOptions.ts [hunk 10]
```diff
@@ -195,6 +196,73 @@ export const globalOptions = {
...
+    prompt: simulationPrompts.defaultChainIdPrompt,
+    validation: z.string(),
+    transform: (chainId: string) => {
+      return chainId as ChainId;
+    },
+    option: new Option(
+      '-c, --default-chain-id <defaultChainId>',
+      'Default chain id for the simulation (eg. 0).',
+    ),
+  }),
```
Comment (alber70g) [inline packages/tools/kadena-cli/src/utils/globalOptions.ts:264]: Maybe we can move the options specifically for simulation to it's own options set.

**Review by alber70g — APPROVED** [review]

File: packages/tools/kadena-cli/src/prompts/simulate.ts
Comment (alber70g) [inline packages/tools/kadena-cli/src/prompts/simulate.ts:1]: Maybe we can move this file to `./src/devnet/prompts/simulate.ts` similar to the options specific to simulation

Comment (nil-amrutlal-dept) [reply packages/tools/kadena-cli/src/prompts/simulate.ts:1]: They are already there. Just being exported from the index.ts` file

File: packages/tools/kadena-cli/src/utils/simulationOptions.ts [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+      '-na, --accounts <accounts>',
+      'Amount of accounts to be created in the simulation.',
+    ),
+  }),
+  simulationTransferInterval: createOption({
+    key: 'interval' as const,
+    prompt: simulationPrompts.transferIntervalPrompt,
+    validation: z.number(),
+    option: new Option(
+      '-ti, --interval <interval>',
```
Comment (nillo) [inline packages/tools/kadena-cli/src/utils/simulationOptions.ts:23]:
```
simulationNumberOfAccounts: createOption({
    key: 'accounts' as const,
    prompt: simulationPrompts.numberOfAccountsPrompt,
    validation: z.number(),
    option: new Option(
      '-na, --accounts <accounts>',
      'Amount of accounts to be created in the simulation.',
    ),
  }),
```

please uphold this format:

```
simulationNumberOfAccounts: createOption({
    key: 'simulationNumberOfAccounts' as const,
    prompt: simulationPrompts.numberOfAccountsPrompt,
    validation: z.number(),
    option: new Option(
      '-na, --simulation-number-of-accounts <simulationNumberOfAccounts>',
      'Amount of accounts to be created in the simulation.',
    ),
  }),
```

For this and all prompts below

File: packages/tools/kadena-cli/src/devnet/utils/simulation/file.ts [hunk 12]
```diff
@@ -0,0 +1,57 @@
...
+export type TransferType =
+  | 'fund'
+  | 'transfer'
+  | 'cross-chain-transfer'
+  | 'safe-transfer'
+  | undefined;
+
+export type TokenActionType = 'create' | 'mint' | 'transfer' | undefined;
+
+export function createDir(directory: string): void {
```
Comment (nillo) [inline packages/tools/kadena-cli/src/devnet/utils/simulation/file.ts:27]: move to filesystem service, so others can use this also, (if not already there)

**Review by nillo — CHANGES_REQUESTED** [review]
lgtm except the minor change request

File: packages/tools/kadena-cli/src/devnet/utils/docker.ts [hunk 13]
```diff
@@ -141,3 +141,13 @@ export function updateDevnet(version?: string): void {
   execSync(`docker pull ${image}`);
   console.log(chalk.green(`Updated ${image}`));
 }
+
+export function containerIsRunning(): boolean {
```
Comment (barthuijgen) [inline packages/tools/kadena-cli/src/devnet/utils/docker.ts:145]: When you use this method, are you making the assumption that the container is exposing a specific port as well? Because this method does not verify it, someone could have started a container using `kadena/devnet` but exposed it on a different port

File: packages/tools/kadena-cli/src/devnet/utils/simulation/coin/simulate.ts [hunk 14]
```diff
@@ -0,0 +1,284 @@
...
+import { safeTransfer } from './safe-transfer.js';
+import { transfer } from './transfer.js';
+
+const simulationTransferOptions: TransferType[] = [
+  'cross-chain-transfer',
+  'transfer',
+  'safe-transfer',
+];
+
+export const MARMALADE_TEMPLATE_FOLDER = 'src/devnet/contracts/marmalade-v2';
```
Comment (barthuijgen) [inline packages/tools/kadena-cli/src/devnet/utils/simulation/coin/simulate.ts:28]: This const isn't being used right? please remove it. (also it wouldn't work if used)

File: packages/tools/kadena-cli/src/devnet/utils/docker.ts (hunk 13 above)
Comment (nil-amrutlal-dept) [reply packages/tools/kadena-cli/src/devnet/utils/docker.ts:145]: I am not using this method anymore. It was indeed not validating the port but the name before. Not it only validates if a container is running. If not useful, I can just remove the function because I am not using it

File: packages/tools/kadena-cli/src/devnet/utils/simulation/coin/simulate.ts [hunk 15]
```diff
@@ -0,0 +1,284 @@
...
+        requestKey: result.reqKey,
+        action: 'fund',
+      });
+    }
+
+    // Generate first seeded random number
+    let seededRandomNo = seedRandom(seed);
+    let counter: number = 0;
+
+    while (true) {
```
Comment (barthuijgen) [inline packages/tools/kadena-cli/src/devnet/utils/simulation/coin/simulate.ts:145]: There is a lot of complexity within this while loop, making it hard for me to see how exactly it will break the loop, or if it is possible that it will run forever. This code would really benefit from moving some logic to separate methods

File: packages/tools/kadena-cli/src/devnet/utils/simulation/file.ts (hunk 12 above)
Comment (barthuijgen) [reply packages/tools/kadena-cli/src/devnet/utils/simulation/file.ts:27]: there is a `ensureDirectoryExists` method

File: packages/tools/kadena-cli/src/devnet/utils/simulation/file.ts [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+    to: '',
+    amount: 0,
+    requestKey: '',
+    action: undefined,
+  };
+  const directory = path.join(`${folder}/`);
+  createDir(directory);
+  const filepath = path.join(directory, filename);
+  const headers = `${Object.keys(exampleData).join(',')}\n`;
+  fs.writeFileSync(filepath, headers);
```
Comment (barthuijgen) [inline packages/tools/kadena-cli/src/devnet/utils/simulation/file.ts:46]: please use the `service.filesystem` instead

File: packages/tools/kadena-cli/src/devnet/utils/docker.ts (hunk 13 above)
Comment (nil-amrutlal-dept) [reply packages/tools/kadena-cli/src/devnet/utils/docker.ts:145]: Removed method since its not being used

File: packages/tools/kadena-cli/src/devnet/utils/simulation/coin/simulate.ts (hunk 14 above)
Comment (nil-amrutlal-dept) [reply packages/tools/kadena-cli/src/devnet/utils/simulation/coin/simulate.ts]: Removed. Well spotted 👍🏽

File: packages/tools/kadena-cli/src/devnet/utils/simulation/file.ts (hunk 12 above)
Comment (nil-amrutlal-dept) [reply packages/tools/kadena-cli/src/devnet/utils/simulation/file.ts:27]: Fixed

File: packages/tools/kadena-cli/src/devnet/utils/simulation/file.ts (hunk 16 above)
Comment (nil-amrutlal-dept) [reply packages/tools/kadena-cli/src/devnet/utils/simulation/file.ts:46]: Fixed

File: packages/tools/kadena-cli/src/devnet/utils/simulation/coin/simulate.ts (hunk 15 above)
Comment (nil-amrutlal-dept) [reply packages/tools/kadena-cli/src/devnet/utils/simulation/coin/simulate.ts:143]: Everything inside the loop is a little bit hard to refactor since I would have to pass a bunch of common variables inside each method and return objects with multiple properties due to the necessary logging in addition to some conditions that validate if the iteration is valid or if it should be skipped.

I can add some more comments to add clarity to what is happening and check if I can refactor it into smaller methods. 
Answering your question: the loop could go for infinity. This loop's goal is the simulate random traffic in the devnet (with the possibility of replicating it via the seed number). The only way to break the loop automatically is with a maximum time flag which is being developed on another PR, otherwise you'd have to shut it yourself.

**Review by barthuijgen — APPROVED** [review]

File: packages/tools/kadena-cli/src/utils/simulationOptions.ts [hunk 17]
```diff
@@ -0,0 +1,69 @@
...
+import type { ChainId } from '@kadena/types';
+import { simulate as simulationPrompts } from '../prompts/index.js';
+export const simulationOptions = {
+  //Simulation
+  simulationNumberOfAccounts: createOption({
+    key: 'simulationNumberOfAccounts' as const,
+    prompt: simulationPrompts.numberOfAccountsPrompt,
+    validation: z.number(),
+    option: new Option(
+      '-na, --simulation-number-of-accounts <simulationNumberOfAccounts>',
```
Comment (nillo) [inline packages/tools/kadena-cli/src/utils/simulationOptions.ts:14]: @alber70g there was some confusion about this
but this should be
-n or only --simulation-number-of-accounts and removal of the -na 

correct | lets get this confusion out :)

**Review by nillo — APPROVED** [review]
Approved but please let Albert confirm my remark.  ;)

File: packages/tools/kadena-cli/src/devnet/commands/devnetSimulation.ts (hunk 7 above)
Comment (alber70g) [reply packages/tools/kadena-cli/src/devnet/commands/devnetSimulation.ts]: Simulation should be network independent. So it can be any network that's running anywhere, not just on `localhost`.

File: packages/tools/kadena-cli/src/utils/simulationOptions.ts (hunk 17 above)
Comment (alber70g) [reply packages/tools/kadena-cli/src/utils/simulationOptions.ts:14]: So since this is a specifically for devnet simulate, should we move it to the `simulate` directory?

Comment (nil-amrutlal-dept) [reply packages/tools/kadena-cli/src/utils/simulationOptions.ts]: Moved to simulation specific directory; updated the flags to single characters
